### PR TITLE
Support for EOSIO

### PR DESCRIPTION
**Pull request by bravoadam**
## Description

Adds support for [EOSIO](eos.io) blockchain. While this is a graphene based blockchain as well, it differs a lot in its implementation from Steem and Bravo and hence most of the code is new.

## Testing instructions

We have added tests in src/tests as well as swift and kotlin tests in their respective directories

## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist

- [ ] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [ ] Update documentation as needed.

### CODE REVIEW

File: coins.json [hunk 1]
```diff
@@ -501,5 +501,16 @@
         "publicKeyType": "secp256k1",
         "transactionPath": "/tx/",
         "explorer": "https://steemblockexplorer.com"
+    },
+    {
+        "id": "eosio",
+        "name": "EOSIO",
```
Comment (hewigovens) [inline coins.json]: just `EOS`

File: coins.json [hunk 2]
```diff
@@ -501,5 +501,16 @@
...
+        "symbol": "EOS",
+        "decimals": 18,
```
Comment (hewigovens) [inline coins.json]: should be 3, same as Steem

File: include/TrustWalletCore/TWEOSKeyType.h [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+#include "TWBase.h"
+
+TW_EXTERN_C_BEGIN
+
+TW_EXPORT_ENUM(uint32_t)
+enum TWEOSKeyType {
+    TWEOSKeyTypeLegacy = 0,
+    TWEOSKeyTypeModernK1 = 1,
+    TWEOSKeyTypeModernR1 = 2,
```
Comment (hewigovens) [inline include/TrustWalletCore/TWEOSKeyType.h]: Did you implement legacy or modern r1? if not, maybe we can remove this enum?

File: src/EOS/Address.cpp [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    for (int i = 0; i < ChecksumSize; ++i) {
+        keyData[PublicKeyDataSize + i] = bytes[i];
+    } 
+}
+
+/// Returns a string representation of the EOS address.
+std::string Address::string() const {
+    size_t b58Size = Base58Size;
+    char b58[b58Size];
+    b58enc(b58, &b58Size, keyData, KeyDataSize);
```
Comment (hewigovens) [inline src/EOS/Address.cpp]: we have a c++ implementation in `Base58.h`, it's easier to use

File: src/EOS/Name.h [hunk 5]
```diff
@@ -0,0 +1,20 @@
+#pragma once
+
+#include "../BinaryCoding.h"
+#include "../Data.h"
+
+using Data = TW::Data;
+
+namespace TW::EOS {
+class Name {
```
Comment (hewigovens) [inline src/EOS/Name.h]: `Name` is a bit general, how about `AccountName`?

File: src/EOS/Address.cpp [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+}
+
+/// Extracts and verifies the key data from a base58 string.
+/// If the second arg is provided, the keyData and isTestNet 
+/// properties of that object are set from the extracted data.
+bool Address::extractKeyData(const std::string& string, Address *address) {
+    // verify if the string has one of the valid prefixes
+    Type type;
+    size_t prefixSize;
+    // std::cout << "Checking prefix for " + string + ".\n";
```
Comment (hewigovens) [inline src/EOS/Address.cpp]: let's remove this test code and legacy support

File: src/EOS/Transaction.cpp [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    buffer.resize(DataSize);
+
+    for(size_t i = 0; i < ChecksumSize; i++) {
+        buffer.push_back(hash[i]);
+    }
+
+    // create a base58 representation of the buffer
+    size_t b58Size = (DataSize + ChecksumSize) * 138 / 100 + 2;;
+    char b58[b58Size];
+    b58enc(b58, &b58Size, buffer.data(), DataSize + ChecksumSize);
```
Comment (hewigovens) [inline src/EOS/Transaction.cpp]: try c++ Base58.h

File: coins.json (hunk 2 above)
Comment (hewigovens) [reply coins.json]: sorry, seems be 4

File: coins.json (hunk 1 above)
Comment (srpatel19590) [reply coins.json]: Okay. will change it to EOS

File: coins.json (hunk 2 above)
Comment (srpatel19590) [reply coins.json]: Agreed. Changing it to 4. However, this code supports all tokens listed on EOS blockchain which can have up to 18 decimals.

File: src/EOS/Address.cpp (hunk 4 above)
Comment (srpatel19590) [reply src/EOS/Address.cpp]: At the time of writing this code Base58 was not available. Updating code to reuse the existing base58.

File: src/EOS/Name.h (hunk 5 above)
Comment (srpatel19590) [reply src/EOS/Name.h]: Not only account name but also all names on EOSIO system (e.g. account name, action name) should follow this convention.

File: src/EOS/Address.cpp (hunk 6 above)
Comment (srpatel19590) [reply src/EOS/Address.cpp]: okay

File: src/EOS/PackedTransaction.h [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace TW::EOS {
+enum class CompressionType {
+    None = 0,
+    // ZLIB = 1,
+};
+
+NLOHMANN_JSON_SERIALIZE_ENUM( CompressionType, {
+{CompressionType::None, "none"},
+// // {CompressionType::ZLIB, "zlib"},
```
Comment (alejandro-isaza) [inline src/EOS/PackedTransaction.h]: why is this commented out?

File: src/EOS/Transaction.h [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    std::vector<Action> contextFreeActions;
+    std::vector<Extension> transactionExtensions;
+    std::vector<Signature> signatures;
+
+    Data contextFreeData;
+
+    void setReferenceBlock(const Data& referenceBlockId);
+
+    static const uint32_t ExpirySeconds = 30;
+    // static const uint32_t ExpirySeconds = 60 * 10;
```
Comment (alejandro-isaza) [inline src/EOS/Transaction.h]: Remove please

File: swift/Tests/Blockchains/EOSTests.swift [hunk 10]
```diff
@@ -0,0 +1,97 @@
+//
+//  EOSSignerTests.swift
+//  TrustWalletCoreTests
+//
+//  Created by Mohak Shah on 26/04/19.
+//
```
Comment (alejandro-isaza) [inline swift/Tests/Blockchains/EOSTests.swift]: Please use the standard file header.

File: src/EOS/PackedTransaction.h (hunk 8 above)
Comment (srpatel19590) [reply src/EOS/PackedTransaction.h]: The serialized transaction can be compressed but not mandatory. EOS has default compression none. We kept it commented to implement in future or when the trustwallet will have zlib support. However, we will remove this for now.

**Comment by bravoadam** [comment]
@alejandro-isaza Can you please merge the PR if everything is okay? We will raise for Bitshare after this merge.

**Pull request merged**